binary_tree.rb

```diff
@@ -0,0 +1,65 @@
+require 'bundler'
+
+Bundler.require(:default)
+
+class BinaryTree
+  attr_reader :left
+  attr_reader :right
+  attr_reader :value
+
+  def initialize(left, right, value)
+    @left = left
+    @right = right
+    @value = value.to_s
+  end
+
+  def self.create(array)
+
+    array.sort!
```

Comment:
Po pierwszym wywołaniu metody `create` każde kolejne używa już posortowanej tablicy, a sortujemy przy każdym wywołaniu tej metody, co dodaje złożoności obliczeniowej.

Jeśli chodzi o sam wybór struktury danych to optymalniejsza do zbalansowania byłaby podrodzina drzew BST - drzewo AVL:
https://medium.com/basecs/the-little-avl-tree-that-could-86a3cae410c7
http://wazniak.mimuw.edu.pl/index.php?title=Algorytmy_i_struktury_danych/S%C5%82owniki#Drzewa_AVL
https://secweb.cs.odu.edu/~zeil/cs361/web/website/Lectures/avl/pages/avlcomplexity.html
Dla sortowania tablicy średnia wydajność to O(n log n), natomiast realnie jest to przypadek pesymistyczny. W pesymistycznym przypadku umieszczenie elementu w drzewie AVL to O(log n), ale praktycznie złożoność będzie lepsza. Elementów jest n, więc sumarycznie jest to pesymistyczne O(n log n).

Czyli reasumując: pesymistyczne O(log n) AVL jest preferowane nad średnie O(log n) sortowania. 🙂